Better feedback export

chosak opened the pull request:
The current process for exporting website feedback is complicated and manual (CFPB folks, see the "How to download user submitted site feedback" page on our internal wiki). This PR attempts to improve that by providing a one-click download solution from the Wagtail admin.

## Additions

- New management command to export feedback CSVs from the command line.
- New Wagtail admin page to download feedback via the browser, automatically segmenting it into our three segments (Ask CFPB, Buying a House, Other). The form allows the user to specify a date range but defaults to the most recent quarter, which is our typical use case.
- New Wagtail permission to limit feedback downloads to certain user groups.
- New documentation page describing all of this new functionality (see screenshot below).
- New third-party package requirement ([backports.csv](https://pypi.org/project/backports.csv/)), requiring a `pip install` of requirements.

## Removals

- This change removes the ability to view and edit feedback from the Django admin. This reduces potential exposure to submitted PII as well as the risk of accidental editing or deletion of feedback.

## Testing

Using a Wagtail account with sufficient privileges (say, the admin account), visit "Export Feedback" in Wagtail, and click the button to download some feedback.

Using an account that doesn't have permission (say, an average developer account), you shouldn't see that option in the Wagtail admin sidebar, and trying to visit the page directly should result in a 403 Permission Denied error.

## Screenshots

![image](https://user-images.githubusercontent.com/654645/66671742-b458a280-ec2a-11e9-9890-10e9149355e6.png)

## Todos

- I've staged changes to our internal wiki documentation that point here, and will tag folks from whom I've requested review on an internal ticket with the link.
- It'd be nice to merge this new export menu item with the existing "Export Ask data" item into a common "Exports" submenu, but I'm leaving that for a future task.

## Checklist

- [X] PR has an informative and human-readable title
- [X] Changes are limited to a single goal (no scope creep)
- [X] Code can be automatically merged (no conflicts)
- [X] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [X] Passes all existing automated tests
- [X] Any _change_ in functionality is tested
- [X] New functions are documented (with a description, list of inputs, and expected output)
- [X] Placeholder code is flagged / future todos are captured in comments
- [X] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [X] Project documentation has been updated
- [X] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

cfarm commented:
> It'd be nice to merge this new export menu item with the existing "Export Ask data" item into a common "Exports" submenu, but I'm leaving that for a future task.

YES. let's do it. see #5287 now there are 3 exports happening. #WagtailIsTheNewWordPressYall

niqjohnson commented on cfgov/v1/models/feedback.py:
```diff
@@ -0,0 +1,86 @@
+            'comment',
+            'currently_own',
+            'expect_to_buy',
+            'email',
```
I know these three columns are present in the current Django admin implementation of the feedback downloads, but I *think* they might be old responses that we no longer collect, they're not specified in our current instructions for processing the feedback, and they're pretty much always blank in my experience (except for junk comments where weird formatting forces things to bleed into these columns). I'll check around and see if we're still using them or not.

chosak replied:
@niqjohnson regarding the presence of these columns, as well as the suggestion from @Scotchester below about adding the URL path: while it would be great to revisit the _content_ of the CSVs, we do have some downstream code (not in this repository) that relies on the current columns being there. So I might prefer to leave that work out of this particular change, especially since the presence of these columns don't affect the complexity much.

chosak commented on cfgov/v1/templates/v1/export_feedback.html:
```diff
@@ -0,0 +1,26 @@
+<style>
+button.export-feedback {
+    margin-top: 2em;
+}
```
Done in 5e49753.

chosak commented on cfgov/v1/templates/v1/export_feedback.html:
```diff
@@ -0,0 +1,26 @@
+		</p>
+        <form method="post">
+            {% csrf_token %}
+			{{ form }}
```
Thanks, done in 5e49753.

chosak commented on cfgov/v1/templates/v1/export_feedback.html:
```diff
@@ -0,0 +1,26 @@
+
+{% block extra_css %}
+<style>
+button.export-feedback {
```
Changed to `.export-feedback-btn` per your suggestion offline in 5e49753.

chosak commented on cfgov/v1/templates/v1/export_feedback.html:
```diff
@@ -0,0 +1,26 @@
+
+    <div class="nice-padding">
+        <p>
+            Download a ZIP file containing feedback submitted to the website.
```
I'm on the fence about adding this description here. The documentation already includes a description of what is included, and if someone's on this page then hopefully they've read the docs about how it works. Plus I actually don't really love that we're hardcoding this split the way we are...but if you think it'd add value to the user, I can add it.

Scotchester replied:
That's fair; I retract the suggestion.

chosak commented:
> I strongly favor sticking to UTF8 as our standard encoding for data transfer, and I favor Excel users learning how to open UTF8 CSVs. But if every stakeholder who downloads feedback gets broken characters, the path of least resistance might be to encode the CSVs as windows-1252 so that they open straight into Excel without character assassination.

@higs4281 thanks for the good feedback. I have two thoughts in response.

First, my understanding is that the officially sanctioned way of reviewing website feedback is through the internal tool at the end of the pipeline, that allows for review of _scrubbed_ feedback, without PII. In theory, at least, stakeholders (and developers!) shouldn't be looking at raw CSV files. Part of the motivation behind these changes is to reduce the need for that.

Second, and more generally, if we expect our end users to be looking at any data in Excel, then it seems like the most user-friendly thing we could do is generate our output in Excel format, to avoid the manual step of converting CSVs. One advantage of doing this is that in theory it'd reduce or avoid at least some Unicode character pains, especially those that cause trouble with Python 2/3. I spent a bit of time trying to migrate this PR to Excel output, actually, before retreating in favor of getting it over the finish line.We already include the [openpyxl](https://github.com/cfpb/cfgov-refresh/blob/e8e8336a0e284d5d81c356d853d0183a43ebeabf/requirements/libraries.txt#L27) package that supports this, and use it as part of the research conference functionality [to generate an Excel export of registrants](https://github.com/cfpb/cfgov-refresh/blob/e8e8336a0e284d5d81c356d853d0183a43ebeabf/cfgov/data_research/research_conference.py#L38).

I do agree that it'd be good to have a standard and reused way of doing exports like this.

chosak commented:
I believe that I've addressed all of the feedback above. I made some changes in 1b6f185 to address the text encoding issue that @Scotchester encountered so will leave this open pending his re-review.

All Travis tests pass; I'm not quite sure what Codecov is complaining about.

higs4281 commented:
> I have two thoughts in response.

@chosak Thank you for the explanations. I think you're taking the right path.

Scotchester commented on docs/website-feedback.md:
````diff
@@ -0,0 +1,99 @@
+specified, use the `--exclude` parameter:
+
+```sh
+cfgov/manage.py export_feedback ask_cfpb --exclude --filename everything_except_ask.csv
````
This now works, but I would like the command to have `--exclude` first, which seems more clear to me.

```suggestion
cfgov/manage.py export_feedback --exclude ask-cfpb --filename everything_except_ask.csv
```

Also, `ask_cfpb` is not a valid slug; it should be `ask-cfpb`. Updated in my suggestion on this line, but please update in all other examples.

chosak replied:
Addressed in 60a0d23.
